File: src/api/java/FuncInterp.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    FuncInterp.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/ 

package com.microsoft.z3;

/**
 * A function interpretation is represented as a finite map and an 'else' value.
 * Each entry in the finite map represents the value of a function given a set
 * of arguments.
 **/
public class FuncInterp extends Z3Object {

    /**
     * An Entry object represents an element in the finite map used to encode a
     * function interpretation.
     **/
    public static class Entry extends Z3Object {

        /**
         * Return the (symbolic) value of this entry.
         *
         * @throws Z3Exception
         * @throws Z3Exception on error
         **/
        public Expr getValue()
        {
            return Expr.create(getContext(),
                    Native.funcEntryGetValue(getContext().nCtx(), getNativeObject()));
        }

        /**
         * The number of arguments of the entry.
         * @throws Z3Exception on error
         **/
        public int getNumArgs()
        {
            return Native.funcEntryGetNumArgs(getContext().nCtx(), getNativeObject());
        }

        /**
         * The arguments of the function entry.
         *
         * @throws Z3Exception
         * @throws Z3Exception on error
         **/
        public Expr[] getArgs()
        {
            int n = getNumArgs();
            Expr[] res = new Expr[n];
            for (int i = 0; i < n; i++)
                res[i] = Expr.create(getContext(), Native.funcEntryGetArg(
                        getContext().nCtx(), getNativeObject(), i));
            return res;
        }

        /**
         * A string representation of the function entry.
         **/
        @Override
        public String toString()
        {
            int n = getNumArgs();
            String res = "[";
            Expr[] args = getArgs();
            for (int i = 0; i < n; i++)
                res += args[i] + ", ";
            return res + getValue() + "]";
        }

        Entry(Context ctx, long obj) {
            super(ctx, obj);
        }

        @Override
        void incRef() {
            Native.funcEntryIncRef(getContext().nCtx(), getNativeObject());
        }

        @Override
        void addToReferenceQueue() {
            getContext().getFuncEntryDRQ().storeReference(getContext(), this);
        }
    }

    /**
     * The number of entries in the function interpretation.
     * @throws Z3Exception on error
     * @return an int
     **/
    public int getNumEntries()
    {
        return Native.funcInterpGetNumEntries(getContext().nCtx(), getNativeObject());
    }

    /**
     * The entries in the function interpretation
     * 
     * @throws Z3Exception
     * @throws Z3Exception on error
     **/
    public Entry[] getEntries()
    {
        int n = getNumEntries();
        Entry[] res = new Entry[n];
        for (int i = 0; i < n; i++)
            res[i] = new Entry(getContext(), Native.funcInterpGetEntry(getContext()
                    .nCtx(), getNativeObject(), i));
        return res;
    }

    /**
     * The (symbolic) `else' value of the function interpretation.
     * 
     * @throws Z3Exception
     * @throws Z3Exception on error
     * @return an Expr
     **/
    public Expr getElse()
    {
        return Expr.create(getContext(),
                Native.funcInterpGetElse(getContext().nCtx(), getNativeObject()));
    }

    /**
     * The arity of the function interpretation
     * @throws Z3Exception on error
     * @return an int
     **/
    public int getArity()
    {
        return Native.funcInterpGetArity(getContext().nCtx(), getNativeObject());
    }

    /**
     * A string representation of the function interpretation.
     **/
    public String toString()
    {
        String res = "";
        res += "[";
        for (Entry e : getEntries())
        {
            int n = e.getNumArgs();
            if (n > 1)
                res += "[";
            Expr[] args = e.getArgs();
            for (int i = 0; i < n; i++)
            {
                if (i != 0)
                    res += ", ";
                res += args[i];
            }
            if (n > 1)
                res += "]";
            res += " -> " + e.getValue() + ", ";
        }
        res += "else -> " + getElse();
        res += "]";
        return res;
    }

    FuncInterp(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    @Override
    void incRef() {
        Native.funcInterpIncRef(getContext().nCtx(), getNativeObject());
    }

    @Override
    void addToReferenceQueue() {
        getContext().getFuncInterpDRQ().storeReference(getContext(), this);
    }
}

File: src/api/dotnet/FuncInterp.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    FuncInterp.cs

Abstract:

    Z3 Managed API: Function Interpretations

Author:

    Christoph Wintersteiger (cwinter) 2012-03-21

Notes:
    
--*/

using System.Diagnostics;
using System;

namespace Microsoft.Z3
{
    /// <summary>
    ///  A function interpretation is represented as a finite map and an 'else' value.
    ///  Each entry in the finite map represents the value of a function given a set of arguments.  
    /// </summary>
    public class FuncInterp : Z3Object
    {
        /// <summary>
        /// An Entry object represents an element in the finite map used to encode
        /// a function interpretation.
        /// </summary>
        public class Entry : Z3Object
        {
            /// <summary>
            /// Return the (symbolic) value of this entry.
            /// </summary>
            public Expr Value
            {
                get
                {
                    return Expr.Create(Context, Native.Z3_func_entry_get_value(Context.nCtx, NativeObject));
                }
            }

            /// <summary>
            /// The number of arguments of the entry.
            /// </summary>
            public uint NumArgs
            {
                get { return Native.Z3_func_entry_get_num_args(Context.nCtx, NativeObject); }
            }

            /// <summary>
            /// The arguments of the function entry.
            /// </summary>
            public Expr[] Args
            {
                get
                {

                    uint n = NumArgs;
                    Expr[] res = new Expr[n];
                    for (uint i = 0; i < n; i++)
                        res[i] = Expr.Create(Context, Native.Z3_func_entry_get_arg(Context.nCtx, NativeObject, i));
                    return res;
                }
            }

            /// <summary>
            /// A string representation of the function entry.
            /// </summary>
            public override string ToString()
            {
                uint n = NumArgs;
                string res = "[";
                Expr[] args = Args;
                for (uint i = 0; i < n; i++)
                    res += args[i] + ", ";
                return res + Value + "]";
            }

            #region Internal
            internal Entry(Context ctx, IntPtr obj) : base(ctx, obj) { Debug.Assert(ctx != null); }

            internal class DecRefQueue : IDecRefQueue
            {
                public DecRefQueue() : base() { }
                public DecRefQueue(uint move_limit) : base(move_limit) { }
                internal override void IncRef(Context ctx, IntPtr obj)
                {
                    Native.Z3_func_entry_inc_ref(ctx.nCtx, obj);
                }

                internal override void DecRef(Context ctx, IntPtr obj)
                {
                    Native.Z3_func_entry_dec_ref(ctx.nCtx, obj);
                }
            };

            internal override void IncRef(IntPtr o)
            {
                Context.FuncEntry_DRQ.IncAndClear(Context, o);
                base.IncRef(o);
            }

            internal override void DecRef(IntPtr o)
            {
                Context.FuncEntry_DRQ.Add(o);
                base.DecRef(o);
            }
            #endregion
        };

        /// <summary>
        /// The number of entries in the function interpretation.
        /// </summary>
        public uint NumEntries
        {
            get { return Native.Z3_func_interp_get_num_entries(Context.nCtx, NativeObject); }
        }

        /// <summary>
        /// The entries in the function interpretation
        /// </summary>
        public Entry[] Entries
        {
            get
            {

                uint n = NumEntries;
                Entry[] res = new Entry[n];
                for (uint i = 0; i < n; i++)
                    res[i] = new Entry(Context, Native.Z3_func_interp_get_entry(Context.nCtx, NativeObject, i));
                return res;
            }
        }

        /// <summary>
        /// The (symbolic) `else' value of the function interpretation.
        /// </summary>
        public Expr Else
        {
            get
            {

                return Expr.Create(Context, Native.Z3_func_interp_get_else(Context.nCtx, NativeObject));
            }
        }

        /// <summary>
        /// The arity of the function interpretation
        /// </summary>
        public uint Arity
        {
            get { return Native.Z3_func_interp_get_arity(Context.nCtx, NativeObject); }
        }

        /// <summary>
        /// A string representation of the function interpretation.
        /// </summary>    
        public override string ToString()
        {
            string res = "";
            res += "[";
            foreach (Entry e in Entries)
            {
                uint n = e.NumArgs;
                if (n > 1) res += "[";
                Expr[] args = e.Args;
                for (uint i = 0; i < n; i++)
                {
                    if (i != 0) res += ", ";
                    res += args[i];
                }
                if (n > 1) res += "]";
                res += " -> " + e.Value + ", ";
            }
            res += "else -> " + Else;
            res += "]";
            return res;
        }

        #region Internal
        internal FuncInterp(Context ctx, IntPtr obj)
            : base(ctx, obj)
        {
            Debug.Assert(ctx != null);
        }

        internal class DecRefQueue : IDecRefQueue
        {
            public DecRefQueue() : base() { }
            public DecRefQueue(uint move_limit) : base(move_limit) { }
            internal override void IncRef(Context ctx, IntPtr obj)
            {
                Native.Z3_func_interp_inc_ref(ctx.nCtx, obj);
            }

            internal override void DecRef(Context ctx, IntPtr obj)
            {
                Native.Z3_func_interp_dec_ref(ctx.nCtx, obj);
            }
        };

        internal override void IncRef(IntPtr o)
        {
            Context.FuncInterp_DRQ.IncAndClear(Context, o);
            base.IncRef(o);
        }

        internal override void DecRef(IntPtr o)
        {
            Context.FuncInterp_DRQ.Add(o);
            base.DecRef(o);
        }
        #endregion
    }
}
